src/validation.cpp

```diff
@@ -1723,7 +1723,32 @@ DisconnectResult CChainState::DisconnectBlock(const CBlock& block, const CBlockI
     view.SetBestBlock(pindex->pprev->GetBlockHash());
 
     if (!fIsFakeNet) {
-        mnview.DecrementMintedBy(pindex->minterKey());
+        const CKeyID minterKey = pindex->minterKey();
+        mnview.DecrementMintedBy(minterKey);
+
+        // Default time to be set if no previous staked block found.
+        int64_t blockTime{-1};
+
+        // Find last block staked by this node.
+        auto nodeId = pcustomcsview->GetMasternodeIdByOperator(minterKey);
+        for (const CBlockIndex* tip = pindex->pprev; nodeId && tip &&
+             pindex->pprev->GetBlockTime() - tip->GetBlockTime() < Params().GetConsensus().pos.nStakeMaxAge &&
+                tip->nHeight >= Params().GetConsensus().DakotaCrescentHeight; tip = tip->pprev)
+        {
+            CKeyID staker;
+            if (tip->GetBlockHeader().ExtractMinterKey(staker))
+            {
+                auto id = pcustomcsview->GetMasternodeIdByOperator(staker);
+                if (id && id == nodeId)
+                {
+                    // Found the last staked block
+                    blockTime = tip->GetBlockTime();
+                    break;
+                }
+            }
+        }
```

Comment:
Why here don't use `GetMasternodeLastBlockTime` ?

Reply:
GetMasternodeLastBlockTime only stores time against a masternode ID so gets overwritten, the entry in GetMasternodeLastBlockTime is the one we are overwriting here.

Reply:
I'm still concern about, disconnect block time will increase.

Reply:
Yes, maybe can consider cache the results.

Reply:
@bvbfan your suggestion was to use <masternode, height>, won't we then be iterating through all records each time to find the latest height on sync for each block, rather than just get the latest time stored against the masternode?

Or do we store latest time against masternode as is with the addition of another <masternode, height> record to iterate over when running DisconnectBlock?

We may still end up iterating over a lot of records to find the latest height, max time for the loop above is 1 second if it has to go all the way back through 7 days of blocks on my computer.

Reply:
@Bushstar yep, that's right. We can get masternode id then you can search by partial key `{masternodeid, 0}` you can find last one.

Reply:
We now store the masternode and time for every single block so the above code is now obsolete.